[v3.6-branch] cmake: modules: extensions: Fix missing board revision overlays

nordicjm opened the pull request:
Fixes an issue whereby board overlays for board revisions were not included when configuring applications

Fixes #71203

57300 committed:
cmake: configuration_files: Add missing FILE_SUFFIX handling

for Kconfig fragments in the `boards` application subdirectory.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
(cherry picked from commit 3cff550180fe83849d6029c93865e088df361617)

nordicjm committed:
cmake: modules: extensions: Fix missing board revision overlays

Fixes an issue whereby board overlays for board revisions were not
included when configuring applications

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on cmake/modules/configuration_files.cmake:
```diff
@@ -77,8 +77,15 @@ zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS APP_BOARD_DTS SUFFIX
-              NAMES "${APP_BOARD_DTS};${BOARD}.overlay;app.overlay" SUFFIX ${FILE_SUFFIX})
+  foreach(file ${APP_BOARD_DTS})
+    zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
+                NAMES "${file}" SUFFIX ${FILE_SUFFIX})
```
Suffix is already applied here: https://github.com/zephyrproject-rtos/zephyr/blob/719b3380555e68843efa54073a256dd62c231751/cmake/modules/configuration_files.cmake#L76

So I assume this code will duplicate `SUFFIX` when used.

nordicjm replied:
Noticed other issues from this, have applied fixes

nordicjm committed:
cmake: modules: extensions: Rename prefixes in functions

Renames prefixes in some functions to avoid clashes with global
variables that have the same name

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2481,76 +2481,76 @@ Please provide one of following: APPLICATION_ROOT, CONF_FILES")
-  cmake_parse_arguments(FILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
-  if(FILE_UNPARSED_ARGUMENTS)
-      message(FATAL_ERROR "zephyr_file(${ARGV0} <val> ...) given unknown arguments: ${FILE_UNPARSED_ARGUMENTS}")
+  cmake_parse_arguments(ZFILE "${options}" "${single_args}" "${multi_args}" ${ARGN})
```
Instead of creating a new commit with combination of multiple commits, then I believe the proper way is to actually do a cherry-pick of needed commits from `zephyr/main`.

That will both make it easier to identify identical changes / commits between branches, as well as help identifying cases where some commits might have been forgotten.

In this case it appears to me that the changes in 6ebb408921c535c45ca4c31b9ded672a0dae456c lacks the fixes from 3cff550180fe83849d6029c93865e088df361617, and hence while trying to fix #71203 you introduce a new error.

Please cherry-pick relevant commits instead, which seems to be the following:
50f0454fbd446b5bb0653b25cedccd022aa18006
3cff550180fe83849d6029c93865e088df361617
030130573a035f50f0cc931763e4767dd3e8ab3b (from PR #71280)

nordicjm replied:
It is not cherry-picked because it does not apply, the changes in those linked commits use MERGE functions which are not in 3.6

tejlmand replied:
I know it's not cherry-picked but you can still cherry-pick and during conflict resolving you'll get the chance to adjust the commit wrt. the MERGE functionality and the board qualifiers.

And the `-x` makes it even easier to see the origin of the fix.

End result in code should of course be identical to re-implementing the changes, but making it much easier to trace which fixes from main that has been applied in the branch.

tejlmand commented on cmake/modules/configuration_files.cmake:
```diff
@@ -77,8 +77,15 @@ zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS APP_BOARD_DTS SUFFIX
 if(NOT DEFINED DTC_OVERLAY_FILE)
-  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
-              NAMES "${APP_BOARD_DTS};${BOARD}.overlay;app.overlay" SUFFIX ${FILE_SUFFIX})
+  foreach(file ${APP_BOARD_DTS})
+    zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
+                NAMES "${file}")
+  endforeach()
```
`zephyr_file()` call here: https://github.com/zephyrproject-rtos/zephyr/blob/719b3380555e68843efa54073a256dd62c231751/cmake/modules/configuration_files.cmake#L76
has already checked for existence of files as well as created a list of files when revision is used.

Thus there is no need for an extra `zephyr_file()` call for each file, just check if `APP_BOARD_DTS` is defined, like this:
```
if(NOT DEFINED DTC_OVERLAY_FILE)
  if(DEFINED APP_BOARD_DTS)
    set(DTC_OVERLAY_FILE ${APP_BOARD_DTS})
  else()
     zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
                NAMES "${BOARD}.overlay;app.overlay" SUFFIX ${FILE_SUFFIX})
  endif()
endif()
```

You can even move the code in L76 inside the if-statement, like this:
```
if(NOT DEFINED DTC_OVERLAY_FILE)
  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards DTS DTC_OVERLAY_FILE SUFFIX ${FILE_SUFFIX})

  if(NOT DEFINED DTC_OVERLAY_FILE)
     zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR} DTS DTC_OVERLAY_FILE
                 NAMES "${BOARD}.overlay;app.overlay" SUFFIX ${FILE_SUFFIX})
  endif()
endif()
```
It's current location probably made sense a long time ago, but not anymore afaict.

nordicjm replied:
Fixed

tejlmand reviewed (approved):
Approved, because code fixes the issue, but would still prefer cherry-picked commits with conflict resolving for trace-ability reasons.

nordicjm commented:
Ping @henrikbrixandersen
